File: 209.minSubArrayLen.go
package main

func minSubArrayLen(target int, nums []int) int {
	left := 0
	right := 0
	size := len(nums)

	min_length := size + 1

	sum := 0

	for left < size {
		for right < size && sum < target {
			sum += nums[right]
			right++
		}

		if sum >= target {
			min_length = min(min_length, right-left)
		}

		sum -= nums[left]
		left++
	}

	if min_length > size {
		return 0
	}

	return min_length
}

File: 209.minSubArrayLen.c
#include "shdstd.h"

#define MIN(x, y) (((x) < (y)) ? (x) : (y))
#define INF64 (INT64_MAX / 4)
int minSubArrayLen(int target, int *nums, int size)
{
	int left, right;
	left = right = 0;

	int64_t min_length = INF64;

	int64_t sum = 0;
	while (left < size) {
		while (right < size && sum < target) {
			sum += (int64_t)nums[right++];
		}

		if (sum >= target) {
			min_length = MIN(min_length, right - left);
		}

		sum -= nums[left++];
	}

	if (min_length == INF64) {
		return 0;
	}

	return min_length;
}
